File: androidApp/src/main/java/com/commuto/interfacemobile/android/offer/OfferTruthSource.kt
package com.commuto.interfacemobile.android.offer

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.snapshots.SnapshotStateMap
import java.math.BigInteger
import java.util.*

/**
 * An interface that a class must implement in order to act as a single source of truth for offer-related data.
 * @property offers A [SnapshotStateMap] mapping [UUID]s to [Offer]s that should act as a single source of truth for all
 * offer-related data.
 * @property serviceFeeRate The current
 * [service fee rate](https://github.com/jimmyneutront/commuto-whitepaper/blob/main/commuto-whitepaper.txt) as a
 * percentage times 100, or `null` if the current service fee rate is not known.
 */
interface OfferTruthSource {
    var offers: SnapshotStateMap<UUID, Offer>
    var serviceFeeRate: MutableState<BigInteger?>

    /**
     * Should add a new [Offer] to [offers].
     *
     * @param offer The new [Offer] that should be added to [offers].
     */
    fun addOffer(offer: Offer)

    /**
     * Should remove from [offers] all [Offer]s with an ID equal to [id]. There should only be one such [Offer].
     *
     * @param id The ID of the [Offer] to remove.
     */
    fun removeOffer(id: UUID)
}

File: iosApp/iosApp/offer/OfferTruthSource.swift
//
//  OfferTruthSource.swift
//  iosApp
//
//  Created by jimmyt on 6/24/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import BigInt
import Foundation

/**
 A protocol that a structure or class must adopt in order to act as a single source of truth for open-offer-related data.
 */
protocol OfferTruthSource: ObservableObject {
    /**
     A dictionary mapping offer IDs (as `UUID`s) to `Offer`s, which is the single source of truth for all open-offer-related data.
     */
    var offers: [UUID: Offer] { get set }
    /**
     The current [service fee rate](https://github.com/jimmyneutront/commuto-whitepaper/blob/main/commuto-whitepaper.txt) as a percentage times 100, or `nil` if the current service fee rate is not known.
     */
    var serviceFeeRate: BigUInt? { get set }
    /**
     Indicates whether this is currently getting the current service fee rate.
     */
    var isGettingServiceFeeRate: Bool { get set }
    /**
     Attempts to get the current service fee rate and set `serviceFeeRate` equal to the result.
     */
    func updateServiceFeeRate()
}
